src/mumble/PipeWire.cpp

```diff
@@ -0,0 +1,437 @@
+			*reinterpret_cast< void ** >(&var) = static_cast< void * >(m_lib.resolve(#var)); \
+			if (!var)                                                                        \
+				return;                                                                      \
+		}
+#endif
+
+static std::unique_ptr< PipeWireSystem > pws;
+
+class PipeWireInputRegistrar : public AudioInputRegistrar {
+private:
+	AudioInput *create() override;
+	const QList< audioDevice > getDeviceChoices() override;
+	void setDeviceChoice(const QVariant &, Settings &) override;
+	bool canEcho(EchoCancelOptionID, const QString &) const override;
+	bool isMicrophoneAccessDeniedByOS() override;
+
+public:
+	PipeWireInputRegistrar();
+};
+
+class PipeWireOutputRegistrar : public AudioOutputRegistrar {
+private:
+	AudioOutput *create() override;
+	const QList< audioDevice > getDeviceChoices() override;
+	void setDeviceChoice(const QVariant &, Settings &) override;
+	bool usesOutputDelay() const override;
+
+public:
+	PipeWireOutputRegistrar();
+};
+
+class PipeWireInit : public DeferInit {
+private:
+	std::unique_ptr< PipeWireInputRegistrar > inputRegistrar;
+	std::unique_ptr< PipeWireOutputRegistrar > outputRegistrar;
+	void initialize() override;
+	void destroy() override;
+};
+
+PipeWireInputRegistrar::PipeWireInputRegistrar() : AudioInputRegistrar(QStringLiteral("PipeWire"), 100) {
+}
+
+AudioInput *PipeWireInputRegistrar::create() {
+	return new PipeWireInput;
+}
+
+const QList< audioDevice > PipeWireInputRegistrar::getDeviceChoices() {
+	return { audioDevice(QStringLiteral("Mono"), QString::number(1)) };
+}
+
+void PipeWireInputRegistrar::setDeviceChoice(const QVariant &choice, Settings &settings) {
+	settings.qsPipeWireInput = choice.toString();
+}
+
+bool PipeWireInputRegistrar::canEcho(EchoCancelOptionID, const QString &) const {
+	return false;
+}
+
+bool PipeWireInputRegistrar::isMicrophoneAccessDeniedByOS() {
+	return false;
+}
+
+PipeWireOutputRegistrar::PipeWireOutputRegistrar() : AudioOutputRegistrar(QStringLiteral("PipeWire"), 100) {
+}
+
+AudioOutput *PipeWireOutputRegistrar::create() {
+	return new PipeWireOutput;
+}
+
+const QList< audioDevice > PipeWireOutputRegistrar::getDeviceChoices() {
+	return { audioDevice(QStringLiteral("Mono"), QString::number(1)),
+			 audioDevice(QStringLiteral("Stereo"), QString::number(2)),
+			 audioDevice(QStringLiteral("2.1"), QString::number(3)),
+			 audioDevice(QStringLiteral("3.1"), QString::number(4)),
+			 audioDevice(QStringLiteral("5.1"), QString::number(6)),
+			 audioDevice(QStringLiteral("7.1"), QString::number(8)) };
+}
+
+void PipeWireOutputRegistrar::setDeviceChoice(const QVariant &choice, Settings &settings) {
+	settings.qsPipeWireOutput = choice.toString();
+}
+
+bool PipeWireOutputRegistrar::usesOutputDelay() const {
+	return false;
+}
+
+void PipeWireInit::initialize() {
+	pws.reset(new PipeWireSystem());
+
+	pws->m_lock.lock();
+	pws->m_waiter.wait(&pws->m_lock, 1000);
+	pws->m_lock.unlock();
```

Comment:
What is this for? There should probably be a comment about that in here

Reply:
It's used in all backends, it just waits for the initialization process to complete.